packages/sync/src/PersistentStore.ts

```diff
@@ -9,3 +8,199 @@ export interface PersistentStore<T> {
+        const cipherText = cipherTextWithIv.slice(cipherTextWithIv.indexOf(":") + 1);
+        const iv = this.strToArray(cipherTextWithIv.substr(0, cipherTextWithIv.indexOf(":")));
+        this.crypto.subtle.decrypt({ name: "AES-GCM", iv },
+          this.masterKey, this.strToArray(cipherText)).then(result => {
+            return this.arrayToStr(result);
+          });
+      }
+    }
+    return retrieved;
+  }
+
+  /**
+   * Stores an ecrypted item in storage. Generates a random initialisation vector for each encryption.
+   * This value is then prepended to the data and also stored for later decryption.
+   * @param key the key to use to store your value
+   * @param data the data to persist
+   */
+  public setItem(key: string, data: PersistedData) {
+    const iv = crypto.getRandomValues(new Uint8Array(12));
+    if (data) {
+      if (this.masterKey) {
+        this.crypto.subtle.encrypt({
+          name: "AES-GCM",
+          iv
+        },
+          this.masterKey,
+          this.strToArray(data.toString())
+        ).then(cipherText => {
+          const stringCipher = this.arrayToStr(cipherText);
+          this.storage.setItem(key, iv + ":" + stringCipher);
+        });
+      } else {
+        this.storage.setItem(key, data);
+      }
+    }
+  }
+
+  /**
+   * Remove an item from storage
+   * @param key the key to use to remove your value
+   */
+  public removeItem(key: string) {
+    this.storage.removeItem(key);
+  }
+
+  private async init() {
+    const masterKey = await this.storage.getItem("masterKey");
+    if (masterKey) {
+      const keyAsString = masterKey.toString();
+      const parsedKey = keyAsString.slice(keyAsString.indexOf(":") + 1);
+      await this.getSalt().then( async (res) => {
+        if (res) {
+          await this.unwrapCryptoKey(this.strToArray(parsedKey), res).then(unwrapped => unwrapped);
+        }
+      });
+    } else {
+      this.generateKey();
+    }
+  }
+
+  private generateKey() {
+    this.crypto.subtle.generateKey(
+      {
+        name: "AES-GCM",
+        length: 256
+      },
+      true,
+      ["encrypt", "decrypt"]
+    ).then(result => {
+      this.masterKey = result;
+      this.wrapCryptoKey(this.masterKey);
+    });
+  }
+
+  // fetches the salt from the stored masterKey and returns it
+  private async getSalt() {
+    const masterKey = await this.storage.getItem("masterKey");
+    if (masterKey) {
+      const keyWithSalt = masterKey.toString();
+      const saltStr = keyWithSalt.substr(0, keyWithSalt.indexOf(":"));
+      // const cipherText = keyWithSalt.slice(keyWithSalt.indexOf(":") + 1);
+      const salt = this.strToArray(saltStr);
+      return salt;
+    }
+  }
+
+  // wrap the master key and store it under the "masterKey" key.
+  // when wrapped it can be stored safely and later unwrapped
+  private async wrapCryptoKey(keyToWrap: CryptoKey) {
+    // get the key encryption key
+    const keyMaterial = await this.getKeyMaterial(this.password);
```

Comment:
I would prefer to have the password value to be passed into this function, will make it easier to test.